#!/bin/sh


################
### MAF = 0% ###
################

mkdir data_maf0

# Remove SNPs with a low MAF frequency.
# A conventional MAF threshold for a regular GWAS is between 0.01 or 0.05, depending on sample size.
# SNPs with a low MAF are rare, therefore power is lacking for detecting SNP‐phenotype associations. These SNPs are also more prone to genotyping errors. The MAF threshold should depend on your sample size, larger samples can use lower MAF thresholds. Respectively, for large (N = 100.000) vs. moderate samples (N = 10000), 0.01 and 0.05 are commonly used as MAF threshold.
#   plink --threads 8 --memory 10000\
#    --bfile data/joint_vqsr_22_220_07\
#    --maf 0\
#    --make-bed\
#    --out data_maf0/joint_vqsr_22_220_08

cp data/joint_22_220_07.bed data_maf0/joint_22_220_08.bed
cp data/joint_22_220_07.bim data_maf0/joint_22_220_08.bim
cp data/joint_22_220_07.fam data_maf0/joint_22_220_08.fam




####################################################
### Step 4 ###

# Delete SNPs which are not in Hardy-Weinberg equilibrium (HWE).
# Check the distribution of HWE p-values of all SNPs.

plink --threads 8 --memory 10000\
 --bfile data_maf0/joint_22_220_08\
 --hardy\
 --out data_maf0/joint_22_220_08

# Selecting SNPs with HWE p-value below 0.00001, required for one of the two plot generated by the next Rscript, allows to zoom in on strongly deviating SNPs. 
#awk '{ if ($9 <0.000001) print $0 }' data_maf0/joint_22_220_08.hwe > data_maf0/joint_22_220_08.zoomhwe.hwe
#awk '{ if ($9 <0.0000000001) print $0 }' data_maf0/joint_22_220_08.hwe > data_maf0/joint_22_220_08.zoomzoomhwe.hwe
Rscript --no-save script/hwe.08.maf0.R  # ggplot2

# By default the --hwe option in plink only filters for controls.
# For binary traits we suggest to exclude: HWE p value <1e?10 in cases and <1e?6 in controls. Less strict case threshold avoids discarding disease‐associated SNPs under selection (see online tutorial at https://github.com/MareesAT/GWA_tutorial/).
# For quantitative traits, we recommend HWE p value <1e‐6.
# Therefore, we use two steps, first we use a stringent HWE threshold for controls, followed by a less stringent threshold for the case data.
plink --threads 8 --memory 10000\
 --bfile data_maf0/joint_22_220_08\
 --hwe 1e-6\
 --make-bed\
 --out data_maf0/joint_22_220_08_filter_step1
#   --hwe: 419536 variants removed due to Hardy-Weinberg exact test.

# The HWE threshold for the cases filters out only SNPs which deviate extremely from HWE. 
# This second HWE step only focusses on cases because in the controls all SNPs with a HWE p-value < hwe 1e-6 were already removed
plink --threads 8 --memory 10000\
 --bfile data_maf0/joint_22_220_08_filter_step1\
 --hwe 1e-10\
 --hwe-all\
 --make-bed\
 --out data_maf0/joint_22_220_09
#   --hwe: 117 variants removed due to Hardy-Weinberg exact test.
#   Note: --hwe-all flag deprecated.  Use "--hwe include-nonctrl".





############################################################
### step 5 ###

# Generate a plot of the distribution of the heterozygosity rate of your subjects.
# And remove individuals with a heterozygosity rate deviating more than 3 sd from the mean.

# Checks for heterozygosity are performed on a set of SNPs which are not highly correlated.
# Therefore, to generate a list of non-(highly)correlated SNPs, we exclude high inversion regions (inversion.txt [High LD regions]) and prune the SNPs using the command --indep-pairwise・
# The parameters ・0 5 0.2・stand respectively for: the window size, the number of SNPs to shift the window at each step, and the multiple correlation coefficient for a SNP being regressed on all other SNPs simultaneously.

plink --threads 8 --memory 10000\
 --bfile data_maf0/joint_22_220_09\
  --exclude /work23/home/nsasa/data/CHM13v2.0_hhv6/gwas/data/inversion.txt --range\
 --indep-pairwise 50 5 0.2\
 --out data_maf0/joint_22_220_09.indepSNP
#   Note: --range flag deprecated.  Use e.g. "--extract range <filename>".
# Note, don't delete the file indepSNP.prune.in, we will use this file in later steps of the tutorial.
#Pruning complete.  11344605 of 22725383 variants removed.

plink --threads 8 --memory 10000\
 --bfile data_maf0/joint_22_220_09\
 --extract data_maf0/joint_22_220_09.indepSNP.prune.in\
 --het\
 --out data_maf0/joint_22_220_09.R_check
# This file contains your pruned data set.

# Plot of the heterozygosity rate distribution
Rscript --no-save script/check_heterozygosity_rate.maf0.R

# The following code generates a list of individuals who deviate more than 3 standard deviations from the heterozygosity rate mean.
# For data manipulation we recommend using UNIX. However, when performing statistical calculations R might be more convenient, hence the use of the Rscript for this step:
Rscript --no-save script/heterozygosity_outliers_list.maf0.R

# Output of the command above: fail-het-qc.txt .
# When using our example data/the HapMap data this list contains 2 individuals (i.e., two individuals have a heterozygosity rate deviating more than 3 SD's from the mean).
# Adapt this file to make it compatible for PLINK, by removing all quotation marks from the file and selecting only the first two columns.
sed 's/"// g' data_maf0/joint_22_220_09.fail-het-qc.txt | awk '{print$1, $2}'> data_maf0/joint_22_220_09.het_fail_ind.txt


# Remove heterozygosity rate outliers.
# --remove data_maf0/joint_22_220_09.het_fail_ind.txt\      # MAF 0.05の除外4例
plink --threads 8 --memory 10000\
 --bfile data_maf0/joint_22_220_09\
 --make-bed\
 --remove data_maf0.05/joint_22_220_09.het_fail_ind.plus.txt\
 --out data_maf0/joint_22_220_10
#   22896852 variants and 238 people pass filters and QC.



############################################################
### step 6 ###

# It is essential to check datasets you analyse for cryptic relatedness.
# Assuming a random population sample we are going to exclude all individuals above the pihat threshold of 0.2 in this tutorial.

# Check for relationships between individuals with a pihat > 0.2.
# --genome --min 0.2\
plink --threads 8 --memory 10000\
 --bfile data_maf0/joint_22_220_10\
 --extract data_maf0/joint_22_220_09.indepSNP.prune.in\
 --genome\
 --out data_maf0/joint_22_220_10.pihat
#   --extract: 8597094 variants remaining.

# The HapMap dataset is known to contain parent-offspring relations. 
# The following commands will visualize specifically these parent-offspring relations, using the z values. 
awk '{ if ($8 >0.9) print $0 }' data_maf0/joint_22_220_10.pihat.genome > data_maf0/joint_22_220_10.zoom_pihat.genome

# Generate a plot to assess the type of relationship.
Rscript --no-save script/Relatedness.maf0.R

# The generated plots show a considerable amount of related individuals (explentation plot; PO = parent-offspring, UN = unrelated individuals) in the Hapmap data, this is expected since the dataset was constructed as such.
# Normally, family based data should be analyzed using specific family based methods. In this tutorial, for demonstrative purposes, we treat the relatedness as cryptic relatedness in a random population sample.
# In this tutorial, we aim to remove all 'relatedness' from our dataset.
# To demonstrate that the majority of the relatedness was due to parent-offspring we only include founders (individuals without parents in the dataset).

plink --threads 8 --memory 10000\
 --bfile data_maf0/joint_22_220_10\
 --filter-founders\
 --make-bed\
 --out data_maf0/joint_22_220_11
#   22896852 variants and 238 people pass filters and QC.

# Now we will look again for individuals with a pihat >0.2.
plink --threads 8 --memory 10000\
 --bfile data_maf0/joint_22_220_11\
 --extract data_maf0/joint_22_220_09.indepSNP.prune.in --genome --min 0.2 --out data_maf0/joint_22_220_11.pihat_min0.2_in_founders
# The file 'pihat_min0.2_in_founders.genome' shows that, after exclusion of all non-founders, only 1 individual pair with a pihat greater than 0.2 remains in the HapMap data.
# This is likely to be a full sib or DZ twin pair based on the Z values. Noteworthy, they were not given the same family identity (FID) in the HapMap data.



### MAF 0.05 のMDSから



# Extract these individuals in HapMap data.
plink --threads 8 --memory 10000\
 --bfile data_maf0/joint_22_220_11\
 --keep data02_maf0.05/ASN_MDS_merge2\
 --make-bed\
 --out data02_maf0/joint_22_220_13
#   --keep: 238 people remaining.
#   22896852 variants and 238 people pass filters and QC.
# Note, since our HapMap data did include any ethnic outliers, no individuls were removed at this step. However, if our data would have included individuals outside of the thresholds we set, then these individuals would have been removed.







###########################################################
### Association analyses ###

# For the association analyses we use the files generated in the previous tutorial (population stratification), named: HapMap_3_r3_13 (with .bed, .bim, and .fam. extensions) and covar_mds.txt


mkdir -p data03_maf0

# MAF 0.05のcov用いて

plink2 --threads 8 --memory 5000\
   --bfile data02_maf0/joint_22_220_13\
   --logistic dominant\
   --covar data02_maf0.05/covar_pca.sex.txt\
   --covar-name sex-PC2\
   --out data03_maf0/plink2_logistic_results_dominant_sex_PC1_PC2

echo -e "CHROM\tPOS\tID\tA1\tA2\tOR\tLOG(OR)_SE\tZ_STAT\tP\tERRCODE" > data03_maf0/plink2_logistic_results_sex_PC1_PC2.PHENO1.glm.logistic.DOMonly.removeNA.tsv
tail -n +2 data03_maf0/plink2_logistic_results_dominant_sex_PC1_PC2.PHENO1.glm.logistic.hybrid | awk '$8=="DOM" && $13!="NA"{print $1"\t"$2"\t"$3"\t"$5"\t"$4"\t"$10"\t"$11"\t"$12"\t"$13"\t"$14}' >> data03_maf0/plink2_logistic_results_sex_PC1_PC2.PHENO1.glm.logistic.DOMonly.removeNA.tsv

Rscript script/ggplot2_manhattan_qq_histogram.R 0 sex_PC1_PC2 DOM



### hhv-6bとのr2

mkdir -p data04_maf0
# chr22
plink2 --threads 8 --memory 5000\
   --bfile data02_maf0/joint_22_220_13\
   --chr 22\
   --make-bed\
   --out data04_maf0/joint_22_220_13_chr22
plink2 --threads 8 --memory 5000\
   --bfile data04_maf0/joint_22_220_13_chr22\
   --recode ped\
   --out data04_maf0/joint_22_220_13_chr22

# eHHV-6B to pseudo snp
cat data04_maf0/joint_22_220_13_chr22.ped | awk '{if($6==1){print $1,$2,$3,$4,$5,$6,"A A"}else{print $1,$2,$3,$4,$5,$6,"G A"}}' > data04_maf0/ped.tmp1
cat data04_maf0/joint_22_220_13_chr22.ped | cut -f7- > data04_maf0/ped.tmp2
paste -d ' ' data04_maf0/ped.tmp1 data04_maf0/ped.tmp2 > data04_maf0/joint_22_220_13_chr22.eHHV6B.ped
echo -e "22\teHHV6B\t0\t1" > data04_maf0/map.tmp
cat data04_maf0/map.tmp data04_maf0/joint_22_220_13_chr22.map > data04_maf0/joint_22_220_13_chr22.eHHV6B.map
rm data04_maf0/map.tmp data04_maf0/ped.tmp1 data04_maf0/ped.tmp2

# ped -> bfile この時変異率0だとA1=0になってしまう。-> その場合r2計算されない -> 0を直しても計算されない！
plink --threads 8 --memory 5000\
 --file data04_maf0/joint_22_220_13_chr22.eHHV6B\
 --make-bed\
 --out data04_maf0/joint_22_220_13_chr22.eHHV6B
##  587719 variants and 238 people pass filters and QC.

head -n 1 data04_maf0/joint_22_220_13_chr22.eHHV6B.bim > data04_maf0/tmp.bim
cat data04_maf0/tmp.bim data04_maf0/joint_22_220_13_chr22.bim > data04_maf0/joint_22_220_13_chr22.eHHV6B.bim
rm data04_maf0/tmp.bim

plink --threads 8 --memory 5000\
 --bfile data04_maf0/joint_22_220_13_chr22.eHHV6B\
 --r2 dprime\
 --ld-snp eHHV6B\
 --ld-window-kb 51400\
 --ld-window 999999\
 --ld-window-r2 0\
 --out data04_maf0/eHHV6B_chr22_LDscore
##  587719 variants loaded from .bim file.
##  wc -l: 567915




### chr2
plink2 --threads 8 --memory 5000\
   --bfile data02_maf0/joint_22_220_13\
   --snp rs1664774871\
   --recode ped\
   --out data04_maf0/joint_22_220_13_chr2SNP

# eHHV-6B to pseudo snp
cat data04_maf0/joint_22_220_13_chr2SNP.ped | awk '{if($6==1){print $1,$2,$3,$4,$5,$6,"A A"}else{print $1,$2,$3,$4,$5,$6,"G A"}}' > data04_maf0/ped.tmp1
cat data04_maf0/joint_22_220_13_chr2SNP.ped | cut -f7- > data04_maf0/ped.tmp2
paste -d ' ' data04_maf0/ped.tmp1 data04_maf0/ped.tmp2 > data04_maf0/joint_22_220_13_chr2SNP.eHHV6B.ped
echo -e "2\teHHV6B\t0\t1" > data04_maf0/map.tmp
cat data04_maf0/map.tmp data04_maf0/joint_22_220_13_chr2SNP.map > data04_maf0/joint_22_220_13_chr2SNP.eHHV6B.map
rm data04_maf0/map.tmp data04_maf0/ped.tmp1 data04_maf0/ped.tmp2

# ped -> bfile この時変異率0だとA1=0になってしまう。-> その場合r2計算されない -> 0を直しても計算されない！
plink --threads 8 --memory 5000\
 --file data04_maf0/joint_22_220_13_chr2SNP.eHHV6B\
 --make-bed\
 --out data04_maf0/joint_22_220_13_chr2SNP.eHHV6B
##  2 variants and 238 people pass filters and QC.

plink --threads 8 --memory 5000\
 --bfile data04_maf0/joint_22_220_13_chr2SNP.eHHV6B\
 --r2 dprime\
 --ld-snp eHHV6B\
 --ld-window-kb 51400\
 --ld-window 999999\
 --ld-window-r2 0\
 --out data04_maf0/eHHV6B_chr2SNP_LDscore
##  587719 variants loaded from .bim file.
##  wc -l: 567915
